### ISSUE TITLE
Allowing nbextensions to be enabled by default

### ISSUE THREAD

**Issue by maartenbreddels** [jupyter/notebook#2824]
I think what many people expect, is that when they install ipywidgets/bqplot/ipyvolume or any other extension that is both python and js for it to 'just work' after installation. (I guess everybody agrees).

What is not possible, is to modify files during the installation process, but it's fine to copy over files. To allow for the default enabling of extensions, lets take ipyvolume as an example, we could put the following in setup.py:
```python
setup( ...
    data_files:         ('etc/jupyter/nbconfig/defaults/', ['ipyvolume.json'])
...)
```
Where the content of that file is:
```json
{
  "load_extensions": {
    "ipyvolume/extension": true
  }
}
```

Then `notebook/nbextensions.py` and `notebook.services.config.manager.ConfigManager` would need modifications to read in `defaults/*.json` from the `read_config_path` etc and update the values from these files (if they are not present).

I have this working locally, where I modified `traitlets.config.manager.BaseJSONConfigManager`, since I can expect this behaviour to be useful in other places. But how it's implemented it another detail.

This is a follow up of #1706, but it's like this change to be as minimal as possible so it's easy to agree on with minimal impact.

**Comment by blink1073** [comment]
Thanks for picking this up, @maartenbreddels!   👍 from me on this approach.

**Comment by hainm** [comment]
>  (I guess everybody agrees).

I agree.

**Comment by takluyver** [comment]
I agree that's the way that involves minimal changes, but I think we should think about it more carefully and make a larger change to produce a more coherent system.

Specifically, I don't think it should be based on installing files in the config system. The more files affect the config, the greater the risk of confusion from the file you're looking at being overridden elsewhere.

What I'd like to see instead is a system where nbextensions install into the nbextensions directory a separate json file with some metadata, including how to load the module. E.g.

```json
{
    "name": "cite2c",
    "description": "Insert citations from Zotero",
    "url": "https://github.com/takluyver/cite2c"
    "load": "cite2c/main"
}
```

Any extensions with these files would be loaded by default, and the user could use config to *disable* them. This has a couple of other benefits:

- It becomes easy to find out what extensions are installed on a system - and therefore to build an interface to conveniently switch them on and off.
- It reduces the likelihood of error messages from extensions which are removed but still configured as enabled.

**Comment by juhasch** [comment]
For the contrib_nbextensions we had `.yaml` files describing each extension for some time, and I think  it has proven to be quite practical. Example:
```
Type: Jupyter Notebook Extension
Compatibility: 4.x, 5.x
Name: Collapsible Headings
Main: main.js
Description: "Allows notebook to have collapsible sections, separated by headings"
Link: readme.md
Icon: icon.png
Parameters:
...
```
Yaml is really so much more human-friendly for non-programmers. 

Also, notebook version compatibility is a constant pain, and should be considered. Many people try to use a years-old nbextension, and with the fast changing API, this nearly never works.

And being able to have an comprehensive overview of available nbextensions that shows a description instead of only a cryptic name is somehow nice, too.

**Comment by maartenbreddels** [comment]
@takluyver 👍  on doing this 'the right' way, I think what I propose is simply a band aid (although it will stop a lot of bleeding), but doesn't solve all of the issues. 
Also @SylvainCorlay pointed [this](https://github.com/jupyter/notebook/issues/1508#issuecomment-224956351) out to me, and is also in favour of doing a more rigorous approach.

@juhasch :+1: on using yaml, and we could have some semver checking, without trying to do a whole package manager, just a warning would that something might break will already be useful.

What I'm afraid of is that it will take at least a year for a big improvement to go out so that people don't feel the pain of enabling widgets or other extensions and all, while this band aid could will give a quick relief.

**Comment by takluyver** [comment]
I personally dislike yaml - it has too many special cases and different possible forms. The main Python implementation I'm aware of also involves C extensions, which makes installation more complicated (pyyaml currently has no wheels for Linux or Mac). JSON is less nice to write, but it's widely understood, simple, and there's a parser for it shipped with Python. I suggested it for these because I don't think those files should need to be human edited that often.

If the consensus is YAML I'll live with it, but for me the drawbacks of YAML outweigh the advantages for this case.

In any case, I think we can do something along these lines for 5.2, especially since @juhasch already has an implementation of basically this. :-)

**Comment by maartenbreddels** [comment]
Let me add that now that widgets are becoming more and more popular I don't want us to lose momentum because of installation issues (I hear those a lot). And most of the issues can be fixed with this (I guess enabling is 80% of the issues) without a major change or giving us headaches when moving on.

**Comment by takluyver** [comment]
@jcb91 I just installed the nbextensions configurator - I think I looked at it some time back, but haven't been using it. It's really slick now! Thanks for your work on it. I think we should bring it into the notebook proper so that everyone sees it. Are you up for that? And do you have a feel for how much work it would take?

I also noticed that the `jupyter_contrib_nbextensions` are installed together as a collection of a load of extensions. I think it makes sense to switch to an approach where installed extensions are loaded unless the user has disabled them - but if we do that, we need to point people to installing extensions individually, not as a giant collection. @jcb91 @juhasch @jfbercher (the three Js of Jupyter extensions ;-) does that sound practical?

**Comment by jcb91** [comment]
> Are you up for that? And do you have a feel for how much work it would take?

Absolutely :) gut feeling would be not a huge amount, since most of the functionality described is already present, with the already-noted exception that we used yaml instead of json. But, I don't see it being a huge amount of effort to add support for json (which I also prefer over yaml in this case, since most end users view the yaml contents through the UI of the configurator anyway), and we could either keep yaml support in parallel, or replace it as desired. The main thing to do first would be to agree on the config file discovery mechanism(s) (currently we attempt to load *all* yaml/yml in nbextensions_path, but it would probably be good to restrict file naming if we go for json) and structure (keys, values, and their uses). Another possibility for json would be to use a subkey of an npm package.json, I suppose.

>  I think it makes sense to switch to an approach where installed extensions are loaded unless the user has disabled them - but if we do that, we need to point people to installing extensions individually, not as a giant collection.

I guess we *could* switch to a bunch of separate repos for each extension, but it'd be quite a lot of repos :fearful: The attraction of the big collection as I see it is that it's a quick way to get a collection of things to try out, which is easy to browse through with the configurator, and it only requires one set of installation machinery (though admittedly this is still a two-step install on non-conda systems). However, I doubt anyone would want *all* of the extensions in the contrib collection enabled by default :laughing: :rocket: :boom:

**Comment by jcb91** [comment]
Also, I've just thought, the other thing that might take a bit of time would be tests; currently the configurator uses a few selenium-based tests for full-stack testing, but I'm not totally _au fait_ with the notebook's own casperjs test framework, so I'm not sure how easy it might be to adapt them / construct alternatives.

And finally, since the contrib repo currently rolls its own install command, it would also be possible to disable anything in the collection not explicitly enabled as part of that custom install, if/when enable-by-default becomes the standard.

**Comment by juhasch** [comment]
As @jcb91 already pointed out, the idea of the current contrib nbextensions repo is that you have a collection that is easy to install and you can simply try out different extensions. This is especially true for extensions that need nbconvert support like preprocessors, exporters or templates.

On the other hand the number of extensions is slowly getting too large for the all-in bundling approach.

Regarding YAML vs. JSON: IMO JSON is good when computers talk to each other, YAML is good when a human "talks" to a computer. I don't think a human should have to count brackets when writing configuration files.

**Comment by bollwyvl** [comment]
👍 extensions be enable-able by default.

+0 to YAML. I love writing YAML (you'd better, if you use conda), and see it as the best editor UI for a JSON-based model... like the "edit metadata" dialog in the notebook would be more usable if it spoke (collapsible, linted, schema'd) YAML. But I do like the portability/inspectability of JSON on disk, despite its diffability handicap. Given that pyzmq, one of the few c upstreams, _still_ creates problems occasionally, i am not clamoring to add it as a hard dependency... even though it's in `conda`!

As [mentioned](https://github.com/jupyter/notebook/issues/2894#issuecomment-333834180) over on #2894, another hat in the ring would be an `entry_point`-based way to put things into the config. For a "pure" javascript nbextension, this python-based approach might not be as useful... but packaging `extension.yaml` and `index.js` is no problem on conda (assuming a recipe someplace) and presumably you can have a data-files-only wheel (assuming a setup.py)? That doesn't seem like much maintainer overhead to let users use the tools they trust for their compute tasks :) `npm` still scares me.

Anyhow, however the baseline config gets on disk, I am very 👍 on there being a user-serviceable,  recoverable, discoverable, beautiful way to manipulate the state of the javascript application for both notebook and lab.

**user-serviceable and recoverable**

I love the CLI, but think something powerful and visual like atom or `configurator` seems really nice. 

At it's simplest: Run configurator, whether it's in the notebook, (or next to the notebook), and you configure the state of the javascript application, reload the page, and you're good. Better, keep the versions around so you can go back to them in case of fail (so don't make the configurator an nbextension, i guess?).

**discoverable**

The package managers don't make this particularly easy (i.e. can't search for `entry_points` or [namespaces](https://packaging.python.org/guides/packaging-namespace-packages/), magic names are weak, etc.), so the "big box of extensions" packages like `contrib` or other/future ones (e.g. genomics, CAD) are valuable for their curation purposes, as if they leverage common licensing, documentation, code review, testing, etc. they are a maintenance win vs many small repos, even if they don't specifically share a common focus.

For the case of, "we'll ship this, but you might not want it", it seems like shipping the assets, but with an in-env baseline would let you have both.
```yaml
load_nbextensions:
    thing-everyone-wants: TRUE
    other-thing-everyone-wants: y
    thing-some-people-want: no
    thing-b: off
```
> The above is a playful illustration of how YAML's "human-editability" can get a little [out of hand](http://yaml.org/type/bool.html). 

Anyhow, back to packaging `contrib`: One could probably apply automation akin to `lerna` to the build of these with wheel/conda, and get a metapackage and _n_ separately-versioned packages, all specified by the version of the "contrib" metapackage... but if the dependencies of all of them are reasonable, that might be a bunch of work for no real gain... unless its for piecemeal upstreaming into a domain-specific pack, as suggested above. Specifically on things that are possible with nbconvert, you can quickly rack up some expensive ones! 

**beautiful**

Perhaps, looking at the metadata discussion above, the right on-disk metadata and documentation format is... the notebook! I think if an extension packed a well-named notebook, with some metadata (visually edited and previewed by the configurator), we'd be able leverage all of the existing machinery and let people make some really nice stuff.

So in `extension.ipynb` (or whatever), use some (visually-defined) tags, `jupyter-extension:name`, `jupyter-extension:screenshots`, etc. Parse the metadata and serve them up as some sandboxed HTML.

**Comment by jcb91** [comment]
For packaging the contrib repo, I think it can still persist through this change to enable-by-default, as long as we update its install script to retain the existing not-enabling behaviour. In order to make the big-bundle paradigm redundant, though, I think we'd need a centralized catalogue to make nbextension packages discoverable over the net. The obvious choice for me would be to search _both_ npm and pypi for specific tags, but then again, perhaps some filtering would be a good idea? Maybe an 'official' github repo that just lists extension URIs? Anyway, this cataloguing is, I think, a separate enough issue that the enable-by-default can proceed without addressing it yet.

> Perhaps, looking at the metadata discussion above, the right on-disk metadata and documentation format is... the notebook! I think if an extension packed a well-named notebook, with some metadata (visually edited and previewed by the configurator), we'd be able leverage all of the existing machinery and let people make some really nice stuff.

Hmm. Interesting idea, you mean the configurator would take over some of the notebook rendering in order to make the metadata attributes editable? This would certainly allow you to mix together arbitrary code for altering/setting metadata with markdown readme-type text, and even examples of the new extension in action. Feels like a fairly extensive bit of work to get it going though, so I'll leave it for now :grimacing: but if it seems like it'd be valuable, I'd be happy to have a stab at it in future.

> Better, keep the versions around so you can go back to them in case of fail (so don't make the configurator an nbextension, i guess?).

This is a neat idea. Currently, jupyter_nbextensions_configurator uses the notebook's javascript config API, which provides no history. The package does already provide serverextensions to add additional handlers, so it would be possible to add a config-with-history API, I guess. In fact, alterations it makes could be stored in a dedicated config file, accessed through the regular config API, I suppose :thinking: I think history is probably beyond the scope of the current issue, but potentially something helpful & interesting to consider for another day.

So, I think I'll proceed as follows, unless anyone thinks of reasons to change:

* adopt versioning for compatibility, as suggested by @maartenbreddels 
* use the configurator's discovery mechanism to find installed extensions, then apply explicit enable/disable statuses, as set in config in the current notebook's manner. This would mean that extensions are 'enabled' (in the sense of loading) by default, but that a custom installer script (e.g. for contrib repo) can still override this by explicitly disabling at install time anything which has not already been explicitly _en_abled

As for the yaml/json decision, how about we:

 * use config stored in a `package.json` file, under a specific key, say (following [this SO suggestion on npm key naming](https://stackoverflow.com/questions/10065564/add-custom-metadata-or-config-to-package-json-is-it-valid)) `org_jupyter.nbextension`. This is simple to implement, and since the configurator's data isn't expected to ever be edited anyway (it sets out only the defaults & types of values stored through the notebook config API), no end-user should really have to do any editing of it anyway, unless they're developing an extension, in which case I think we can fairly assume that they are able to familiarise themselves with the json syntax.
 * additionally support the existing yaml config format *if yaml is importable*. This way, bundles like contrib can add yaml as a dependency without forcing it on notebook as a whole.

**Comment by yuvipanda** [comment]
Did this go anywhere? We have something very similar to the proposed setup for server extensions now.

**Comment by bollwyvl** [comment]
Yeah this shipped in 5.3 with the conf.d stuff. some nbextension have
already moved to it.

**Comment by yuvipanda** [comment]
@bollwyvl sweet! Is it documented someplace?

**Comment by bollwyvl** [comment]
Here's the only spot I know of... is there another place that is misleading?

http://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Distributing%20Jupyter%20Extensions%20as%20Python%20Packages.html#Automatically-enabling-a-server-extension-and-nbextension
